sdk/Sdk.Generators/FunctionMetadataProviderGenerator.cs

```diff
@@ -0,0 +1,841 @@
+                indentedTextWriter.WriteLine("using Microsoft.Azure.Functions.Core;");
+                indentedTextWriter.WriteLine("using Microsoft.Azure.Functions.Worker.Core.FunctionMetadata;");
+                indentedTextWriter.WriteLine("using Microsoft.Extensions.DependencyInjection;");
+                indentedTextWriter.WriteLine("using Microsoft.Extensions.Hosting;");
+
+                // create namespace
+                indentedTextWriter.WriteLine("namespace Microsoft.Azure.Functions.Worker");
+                indentedTextWriter.WriteLine("{");
+                indentedTextWriter.Indent++;
+
+                // create class that implements IFunctionMetadataProvider
+                indentedTextWriter.WriteLine("public class GeneratedFunctionMetadataProvider : IFunctionMetadataProvider");
+                indentedTextWriter.WriteLine("{");
+                indentedTextWriter.Indent++;
+
+                WriteGetFunctionsMetadataAsyncMethod(indentedTextWriter, receiver, compilation);
+                AddEnumTypes(indentedTextWriter);
+
+                indentedTextWriter.Indent--;
+                indentedTextWriter.WriteLine("}");
+
+                // add method that users can call in startup to register the source-generated file
+                AddRegistrationExtension(indentedTextWriter);
+
+                indentedTextWriter.Indent--;
+                indentedTextWriter.WriteLine("}");
+
+                indentedTextWriter.Flush();
+                sourceText = SourceText.From(stringWriter.ToString(), encoding: Encoding.UTF8);
+            }
+            // Add the source code to the compilation
+            context.AddSource($"GeneratedFunctionMetadataProvider.g.cs", sourceText);
+        }
+
+        /// <summary>
+        /// Register a factory that can create our custom syntax receiver
+        /// </summary>
+        /// <param name="context"></param>
+        public void Initialize(GeneratorInitializationContext context)
+        {
+            context.RegisterForSyntaxNotifications(() => new SyntaxReceiver());
+        }
+
+        /// <summary>
+        /// This method populates a dictionary with argumentName (keys) and argumentValue (value) pairs given attribMethodSymbol, which
+        /// represents the attribute constructor as as an IMethodSymbol, and attributeData which stores constructor argument values.
+        /// </summary>
+        /// <param name="attribMethodSymbol">The attribute's constructor as an IMethodSymbol.</param>
+        /// <param name="attributeData">Contains constructor arguments for the constructor represented in attribMethodSymbol.</param>
+        /// <param name="dict">A dicitonary to be populated with constructor arguments.</param>
+        /// <exception cref="InvalidOperationException"></exception>
+        internal static void LoadConstructorArguments(IMethodSymbol attribMethodSymbol, AttributeData attributeData, IDictionary<string, object> dict)
```

Comment:
Is these used external to this class? Can it be `private` if not? Or is there another reason `internal`?